pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -912,16 +906,14 @@ protected WriteInEventLoopCallback newObject(Handle<WriteInEventLoopCallback> ha
     }
 
     private void clearPendingMessagesWhenClose() {
-        PulsarClientException ex = new PulsarClientException.AlreadyClosedException(
-                format("The producer %s of the topic %s was already closed when closing the producers",
-                        producerName, topic));
-        pendingMessages.forEach(msg -> {
-            client.getMemoryLimitController().releaseMemory(msg.uncompressedSize);
-            msg.sendComplete(ex);
-            msg.cmd.release();
-            msg.recycle();
-        });
-        pendingMessages.clear();
+        setState(State.Closed);
+        synchronized (this) {
```

Comment:
in the previous code "setState" was inside the `synchronized` block
so probably is is better to keep it inside that block.

At that point we can simply move the  `synchronized` keyword in the signature of the method

Reply:
I think it's okay to move `setState` outside the `synchronized` block. I think it's a problem of previous code. See `connectionOpened` and `connectionFailed`, `setState` is all called outside the `synchronized` block since it's an atomic operation. It's only called inside the `synchronized` block in `closeAsync`, which is unnecessary.

Reply:
It's not clear to me why the `setState` method was ever in a `synchronized` block. In looking at the git history, it has been that way since the initial import. Since `setSet` has its own concurrency control and since we call `setState` outside of `synchronized` blocks in many other parts of the `ProducerImpl`, I don't see why this update should be in the `synchronized` block. Let me know if you think I am missing something, @eolivelli.

Reply:
I don't remember the original reasoning, though if the variable state is changed within the sync block, it would appear atomic also with all the other changes in the sync block.

Reply:
> it would appear atomic also with all the other changes in the sync block.

Yes, that is true. I'll move it back into the sync block. I wonder if these should be updated too? https://github.com/apache/pulsar/blob/da5bac990042e86d339fc4cba431b555ec293d6d/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L1504-L1517

I don't think it will matter too much because once the state is set to a failure/closed state, the producer won't accept new messages anyway.